servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java

```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
-                // instead we just set to MIN_VALUE again and try to re-acquire the lock in case we raced again.
-                if (n < 0) {
-                    subscriptionRequestQueueUpdater.set(this, Long.MIN_VALUE);
-                } else {
-                    subscriptionRequestQueueUpdater.accumulateAndGet(this, n,
-                            FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-                }
-                if (!subscriptionLockOwnerUpdater.compareAndSet(this, null, currentThread)) {
-                    return;
-                }
-                // We previously added our n contribution to the queue, but now that we have acquired the lock
-                // we are responsible for draining the queue.
-                n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                if (n == 0) {
-                    // It is possible that the previous consumer has released the lock, and drained the queue before we
-                    // acquired the lock and drained the queue. This means we have acquired the lock, but the queue has
-                    // already been drained to 0. We should release the lock, try to drain the queue again, and then
-                    // loop to acquire the lock if there are elements to drain.
-                    subscriptionLockOwner = null;
-                    n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                    if (n == 0) {
-                        return;
-                    } else {
-                        continue;
-                    }
+            pendingDemand = mapInvalidRequestN(n);
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
+                                prevPendingDemand + min(Long.MAX_VALUE - prevPendingDemand, n))) {
+                    break;
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        boolean tryAcquire;
+        do {
+            final long acquireId = tryAcquireReentrantLock(subscriptionLockUpdater, this);
+            if (acquireId == 0) {
+                break;
             }
+
             try {
-                subscription.request(n);
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
```

Comment:
A general characteristics of the ConcurrentUtil locks is if an exception occurs while we hold the lock, we have to successfully release the lock otherwise we may miss events from other threads. In this case that mean we miss a `cancel` which is not desirable. We could assume it is the caller's responsibility to recover/cancel but if the cleanup has already been done from the caller's perspective it may not happen again.

Reply:
Practically, we are discussing the case when `subscription.request()` or `subscription.cancel()` throws which in turn makes this `request()` method throws. 

Once a `Subscription` throws thus violating the spec, any expectation around further interaction with the `Subscription` doing anything meaningful isn't really justified. If this helps us simplify control-flow, I think it is a valid approach to take.

Reply:
> any expectation around further interaction with the Subscription doing anything meaningful isn't really justified.

Agreed in general the RS spec is violated but this may happen in practice and will be challenging to debug if a pending/subsequent `cancel()` is not propagated because an in process `request(..)` throws. The concurrency is introduced by this class and can be encapsulated in the class and IMHO the simplification on control flow is not worth the potential complexity to debug later. The control. flow is also different between `request` and `cancel` so it isn't necessarily the case that an exception/bug in `request` doesn't mean `cancel` also has a bug. Making a best effort to propagate means the problem would be debugged elsewhere and won't require unrolling concurrent access trying to understand when/how signals were dropped.

Reply:
I agree that the control-flow simplification isn't substantial. 

You are correct that the following sequence of events is safe to allow:

```
request(1) <- throws
cancel() <- passes
```

but are the following sequence of events also safe or improve debuggability:

```
request(1) <- throws
request(1) <- passes

now, the Subscriber is expecting 2 items but only will ever get 1
```

Reply:
the second sequence is in the undefined behavior territory. I think reasonable behavior for `ConcurrentSubscription`, which is an intermediate layer, is to deliver signals as it receives them instead of dropping/filtering signals because it doesn't have scope to know if they will be useful or not (e.g. debugging, making progress, etc.). The canonical usage of the `ConcurrentUtil` locks facilitate this (e.g. `while(tryAcquire) { try { /* critical section */ } finally { release } }`) without custom exception handling. Are you suggesting actively preventing scenario 2 but allowing scenario 1, and what would we gain (we are in unexpected error territory)?